src/QsCompiler/Transformations/CallGraphWalker.cs

```diff
@@ -19,6 +18,254 @@ namespace Microsoft.Quantum.QsCompiler.Transformations
 
+    /// <summary>
+    /// Implementation of Johnson's algorithm for finding all simple cycles in a graph.
+    /// A simple cycle is one where no node repeats in the cycle until getting back
+    /// to the beginning of the cycle.
+    ///
+    /// The algorithm uses Tarjan's algorithm for finding all strongly-connected
+    /// components in a graph. A strongly-connected component, or SCC, is a subgraph
+    /// in which all nodes can reach all other nodes in the subgraph.
+    ///
+    /// The algorithm begins by using Tarjan's algorithm to find the SCCs of the graph.
+    /// Each SCC is processed for cycles, using the smallest-id node of the SCC as the
+    /// starting node. Then the starting node is removed from the graph, and Tarjan is used
+    /// again to find SCCs of the new graph, and the next SCC is processed. This repeats
+    /// until there are no more SCCs (and no more nodes) left in the graph.
+    ///
+    /// The algorithm has time complexity of O((n + e)(c + 1)) with n nodes, e edges,
+    /// and c elementary cycles.
+    ///
+    /// The starting point for this algorithm in this class is the GetAllCycles method.
+    /// </summary>
+    internal class JohnsonCycleFind
+    {
+        private Stack<(HashSet<int> SCC, int MinNode)> SccStack = new Stack<(HashSet<int> SCC, int MinNode)>();
+
+        /// <summary>
+        /// Johnson's algorithm for finding all cycles in a graph.
+        /// This returns a list of cycles, each represented as a list of nodes. Nodes
+        /// for this algorithm are integers. The cycles are guaranteed to not contain
+        /// any duplicates and the last node in the list is assumed to be connected
+        /// to the first.
+        /// 
+        /// This implementation passes the full graph along with a hash set of nodes
+        /// to represent subgraphs because it is more performant to check if a node
+        /// is in the hash set than to build a dictionary for the subgraph.
+        /// </summary>
+        public List<List<int>> GetAllCycles(Dictionary<int, List<int>> graph)
+        {
+            var cycles = new List<List<int>>();
+
+            PushSCCsFromGraph(graph, graph.Keys.ToHashSet());
+            while (SccStack.Any())
+            {
+                var (scc, startNode) = SccStack.Pop();
+                cycles.AddRange(GetSccCycles(graph, scc, startNode));
+                scc.Remove(startNode);
+                PushSCCsFromGraph(graph, scc);
+            }
+
+            return cycles;
+        }
+
+        /// <summary>
+        /// Uses Tarjan's algorithm for finding strongly-connected components, or SCCs of a
+        /// graph to get all SCC, orders them by their node with the smallest id, and pushes
+        /// them to the SCC stack.
+        /// </summary>
+        private void PushSCCsFromGraph(Dictionary<int, List<int>> graph, HashSet<int> filter)
+        {
+            var sccs = TarjanSCC(graph, filter).OrderByDescending(x => x.MinNode);
```

Comment:
>var sccs [](start = 12, length = 8)

Please consider using explicit type here as per our recently adopted style guide: "Do not use var when the type is not apparent from the right side of the assignment." See [https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/inside-a-program/coding-conventions](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/inside-a-program/coding-conventions)

Reply:
Addressed in comment below.